fri/src/verifier/mod.rs

```diff
@@ -368,3 +497,83 @@ fn get_query_values<E: FieldElement, const N: usize>(
 
     result
 }
+
+fn iterate_through_query<B, E, H, const N: usize>(
+    layer_commitments: &Vec<H::Digest>,
+    folding_roots: &Vec<B>,
+    layer_alphas: &Vec<E>,
+    query: &Vec<(Vec<<H>::Digest>, [E; N])>,
+    position: usize,
+    num_partitions: usize,
+    folding_factor: usize,
+    number_of_layers: usize,
+    domain_size: usize,
+    evaluation: &E,
+    domain_generator: B,
+    max_degree_plus_1: usize,
+    domain_offset: B,
+) -> Result<(usize, E, usize, usize), VerifierError>
```

Comment:
This looks great! A few comments:

1. Let's get rid of `num_partitions` - we can assume that it is always the same (`1`, I believe).
2. Instead of passing `query` as is now, we should pass `AdviseProvider` struct we've discussed in other places. This struct should provide the following functionality:
  a. `get_node` function as discussed before.
  b. `unhash_node` function which given a node value (`Digest`) returns a set of values which this digest represents (we can probably chat about this).
3.  Do we need to pass `max_degree_plus_1` into this function? It seems like it is not relevant to each query - i.e., we could do the check for `max_degree_puls_1` in a different function just once, rather than for each query.
4. Let's assume that `domain_offset` is hard-coded. So, we can create a constant for it rather than pass it as a parameter.

Reply:
Thank you Bobbin!
I think I have wrapped my mind on what we are trying to achieve.
Now for the implementation, should the `AdviceProvider` be generic over `Digest`? I think this is the most straightforward way as the `AdviceProvider` would need to be constructed inside the `unbatch` method of `VerifierChannel`.

Reply:
> 3.  Do we need to pass `max_degree_plus_1` into this function? It seems like it is not relevant to each query - i.e., we could do the check for `max_degree_puls_1` in a different function just once, rather than for each query.

We don't need it if we remove the `max_degree_plus_1 % N != 0` test.